### PR TITLE
🐛 report lack of webhook side effects

### LINKED ISSUE #2851: kubectl diff for CAPI objects
**User Story**

As a user or operator, I would like to use `kubectl diff` to understand what object changes will result when I run an apply.

**Anything else you would like to add:**

Currently the biggest issue is that our webhooks are not configured to permit dry run:

```
Error from server (BadRequest): admission webhook "default.kubeadmcontrolplane.controlplane.cluster.x-k8s.io" does not support dry run
```

This worked in the v1alpha2 timeframe, but maybe that was so because we didn't have webhooks then?

/kind feature
/assign
/lifecycle active

### PR DESCRIPTION
**Pull request by sethp-nr**
**What this PR does / why we need it**:

Otherwise the API server will refuse to call our webhooks on dry run,
which breaks kubectl diff.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2851

Putting this on hold until https://github.com/kubernetes-sigs/controller-tools/pull/429 goes in and we can drop the `replace`

/hold

### CODE REVIEW

**Comment by vincepri** [comment]
/hold

until upstream functionality is merged

**Comment by vincepri** [comment]
@sethp-nr https://github.com/kubernetes-sigs/controller-tools/releases/tag/v0.2.9

File: hack/tools/go.mod [hunk 1]
```diff
@@ -8,9 +8,15 @@ require (
...
 	sigs.k8s.io/kubebuilder/docs/book/utils v0.0.0-20200226075303-ed8438ec10a4
 	sigs.k8s.io/kustomize/kustomize/v3 v3.5.4
 	sigs.k8s.io/testing_frameworks v0.1.2
 )
+
+replace (
+	k8s.io/api => k8s.io/api v0.17.0
+	k8s.io/apiextensions-apiserver => k8s.io/apiextensions-apiserver v0.17.0
+	k8s.io/apimachinery => k8s.io/apimachinery v0.17.0
+)
```
Comment (vincepri) [inline hack/tools/go.mod:22]: we shouldn't need these anymore

Comment (sethp-nr) [reply hack/tools/go.mod:22]: Good catch – will remove.

**Comment by vincepri** [comment]
LGTM pending squash

**Comment by detiber** [comment]
Should we create followup Issues for this against the various known providers to also implement the same?

**Comment by vincepri** [comment]
Sounds good to. me

**Pull request merged**